### PR TITLE
BUG: Fix alpha for transparent colormaps

### PR DESCRIPTION
**Pull request by larsoner**
For plotting functions, the colorbar has a white background. This is problematic because colormaps with an alpha channel are displayed incorrectly. This sets the background color to mid gray, and thus also sets the `cmaplist` entries that are below `threshold` just to be fully transparent. This should have the same effect as using `(0.5, 0.5, 0.5, 1.0)` directly but is more consistent with what the plotting is actually doing.

### CODE REVIEW

**Comment by jasmainak** [comment]
+1. Thanks @larsoner !

**Comment by bthirion** [comment]
Sounds good, but we need to have CirceCI green to looks at the impact on the images.
Is there any particular image in the doc that we can look at to appreciate the change ?
Thx !

**Comment by larsoner** [comment]
It requires using a colormap with custom alpha in your colormap, like:
```
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from nilearn import plotting
from nilearn import datasets

cmap = plt.get_cmap('viridis')(np.linspace(0, 1, 256))
cmap[:, 3] = np.clip(np.abs(np.linspace(-1, 1, 256)), 0, 1)  # middle should have opacity=0
cmap = ListedColormap(cmap)

motor_images = datasets.fetch_neurovault_motor_task()
stat_img = motor_images.images[0]
display = plotting.plot_stat_map(stat_img, threshold=None, colorbar=True,
                                 cmap=cmap)
```
On `master` the middle becomes white even though it's transparent (wrong):
screenshot from 2018-09-13 16-59-13

On this PR it goes to middle gray (right / consistent with what `threshold` does):
screenshot from 2018-09-13 17-00-14

**Comment by larsoner** [comment]
BTW I saw that CircleCI died but the failures did not look related to my PR. If I'm mistaken and they are, let me know and I'll take another look (though it looks like someone has already restarted Circle).

**Comment by KamalakerDadi** [comment]
Thanks! Your use case seems like a bit advanced like using an customised colormap.

- Could you please update about this in doc/whats_new.rst ?
- Do you think some documentation is missing on this aspect, for instance at the ```cmap``` argument in plotting functions or at plotting guide ?

Just curious (optional): Do you always used transparent colormaps for plotting ? If yes, how do you deal with qualitative colormaps ?

**Comment by larsoner** [comment]
> Your use case seems like a bit advanced like using an customised colormap.

Yes, indeed it is.

> Could you please update about this in doc/whats_new.rst ?

Done, let me know if it's in the right place!

> Do you think some documentation is missing on this aspect, for instance at the cmap argument in plotting functions or at plotting guide ?

No I don't think anything is missing. Basically this does what I would have expected: put a middle gray background on the colorbar and allow it to "show through" when using colormaps with non-unity alpha values. So in that sense it's just a bugfix to obtain what advanced/custom colormap users would expect the behavior to be.

> Just curious (optional): Do you always used transparent colormaps for plotting ? If yes, how do you deal with qualitative colormaps ?

Not always, no. In MNE-Python we give users the option of making the 0->0.5 part of the colormap transition linearly from transparent to opaque with a `transparent` keyword argument. It's basically a soft threshold.

We don't currently have a good way of dealing with qualitative colormaps.

File: doc/whats_new.rst [hunk 1]
```diff
@@ -66,6 +66,9 @@ Changes
     - `sklearn.cross_validation` and `sklearn.grid_search` have been
       replaced by `sklearn.model_selection` in all the examples.
 
+    - Colorbars in plotting functions now have a middle gray background
+      suitable for use with custom colormaps with a non-unity alpha channel.
```
[the remark this answers is not in the document]
Comment (larsoner) [reply doc/whats_new.rst]: You can use it with any activation data including fMRI

[the remark this answers is not in the document]
Comment (larsoner) [reply doc/whats_new.rst]: (really any sort of overlay that makes use of a colormap)

**Comment by KamalakerDadi** [comment]
Views with glass brain with the same approach as above but only the change of plot_stat_map to plot_glass_brain to see how things went with glass brain. The only difference I can see is a colorbar effect showing gray rather than white when colormap is transparent.

This PR
gl
gl_abs_false

Current master:
gl
gl_abs_false


@GaelVaroquaux Do you any objections to merge ?

File: nilearn/plotting/displays.py [hunk 2]
```diff
@@ -775,6 +776,7 @@ def _show_colorbar(self, cmap, norm, threshold=None):
             self._colorbar_ax, ticks=ticks, norm=norm,
             orientation='vertical', cmap=our_cmap, boundaries=bounds,
             spacing='proportional', format='%.2g')
+        self._cbar.patch.set_facecolor(mid_gray)
```
Comment (KamalakerDadi) [inline nilearn/plotting/displays.py]: Is it this enough ? Do you also need adding (0.) ?

For me it looks like using only this ```self._cbar.patch.set_facecolor(mid_gray)``` is sufficient. Correct me if I am wrong.

**Comment by larsoner** [comment]
Actually that looks like a bug was inadventently introduced. `glass_brain` should not use mid gray but instead white for the background. I'll look into it

File: nilearn/plotting/displays.py (hunk 2 above)
Comment (larsoner) [reply nilearn/plotting/displays.py]: This sets the background color, if you add `(0.,)` it makes it transparent and thus defeats the purpose

**Comment by larsoner** [comment]
Running this code now produces (what I would think of as) the expected result:
```
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from nilearn import plotting
from nilearn import datasets

cmap = plt.get_cmap('viridis')(np.linspace(0, 1, 256))
cmap[:, 3] = np.clip(np.abs(np.linspace(-1, 1, 256)), 0, 1)  # middle should have opacity=0
cmap = ListedColormap(cmap)

motor_images = datasets.fetch_neurovault_motor_task()
stat_img = motor_images.images[0]
for black_bg in (True, False):
    display = plotting.plot_stat_map(stat_img, threshold=None, colorbar=True,
                                     cmap=cmap, black_bg=black_bg)
    plotting.plot_glass_brain(stat_img, threshold=None, colorbar=True, cmap=cmap,
                              black_bg=black_bg)
```
screenshot from 2018-10-04 10-24-00
screenshot from 2018-10-04 10-24-01
screenshot from 2018-10-04 10-24-03
screenshot from 2018-10-04 10-24-04

File: nilearn/plotting/displays.py [hunk 3]
```diff
@@ -519,7 +523,7 @@ def init_with_figure(cls, img, threshold=None,
         # People forget to turn their axis off, or to set the zorder, and
```
Comment (kchawla-pi) [inline nilearn/plotting/displays.py]: set the border,

Comment (kchawla-pi) [reply nilearn/plotting/displays.py]: forget to turn their axes off (not axis)

File: nilearn/plotting/displays.py [hunk 4]
```diff
@@ -764,7 +768,7 @@ def _show_colorbar(self, cmap, norm, threshold=None):
         istart = int(norm(-offset, clip=True) * (our_cmap.N - 1))
```
Comment (kchawla-pi) [inline nilearn/plotting/displays.py]: More descriptive names for istart & istop would be helpful. Like why these are the start & stop values. 
A comment explaining what is being done here would be helpful.

**Comment by KamalakerDadi** [comment]
Sorry for being late.

Could you please rebase on top of master so that CircleCI will build to have a look at the plotting gallery. It might also be useful for others to pitch in.

Could you please add tests related to ```brain_color```

+1 from my side.

**Comment by larsoner** [comment]
> Could you please add tests related to brain_color

This is a hidden/private param that differs between `plot_glass_brain` using white and `plot_stat_img` using middle gray. So it should already be implictly tested as long as these functions are called somewhere.

File: nilearn/plotting/displays.py (hunk 3 above)
Comment (larsoner) [reply nilearn/plotting/displays.py]: I think it actually could be related to `zorder` here. And the `mpl` is `axis('off')` in mpl so I think that's okay, too. (I also did not modify this code/comment so maybe such general cleanup would be better in a separate PR.)

**Comment by larsoner** [comment]
Rebased and pushed a commit to change the `istart`/`istop` variables to have more verbose names

File: nilearn/plotting/displays.py (hunk 3 above)
Comment (kchawla-pi) [reply nilearn/plotting/displays.py]: Ah as in z_order.
> I  also did not modify this code/comment so maybe such general cleanup would be better in a separate PR.
Cool.

**Comment by GaelVaroquaux** [comment]
This is so pretty (and useful, as it avoids thresholding). I would push for an option "alpha_gradient" that does this in the various plotting option (are other people on board with me here?).

I love it. I'm merging this, with the hope that the problem on circle will be solved.
